### DISCUSSION TITLE
Console.CursorTop always zero when stdin redirected on linux

### DISCUSSION THREAD

**Discussion by lejsekt** [dotnet/runtime#57347] [Q&A] [answered]
Hi,
I'm hitting the following issue with Console.CursorTop. It seems to report the correct number on Windows regardless of whether the stdin is redirected or not. However, it is always zero on linux when stdin is redirected. What am I doing wrong?

Program.cs of a .NET 5 console app.

```
using System;

namespace cursor_top
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine($"{Console.CursorTop}");
        }
    }
}
```

I'm using Windows Terminal on Windows and run it like the following in pwsh:
```
dotnet run
echo "foo" | dotnet run
```

For linux test env I'm using ubuntu instance in WSL2 and I publish the program as a self contained single file executable. I run it like:
```
./cursor_top
echo "foo" | ./cursor_top
```

The last always shows cursor top as zero.

**Comment by stephentoub** [comment, marked as answer]
There's no formal API on Unix for reading the current cursor position.  Instead, it works by writing a known ANSI escape sequence to stdout, to which a conforming terminal responds by providing the cursor position for reading on stdin.  As such, if either stdout or stdin is redirected, we can't get the cursor position on Unix.
https://github.com/dotnet/runtime/blob/134c25d532db689c63615e9eb13cf31d3fdf0942/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L439-L444

**Reply by lejsekt** [reply]
Thank you for the explanation. Is there a workaround?

I'm actually trying to display progress based on the amount of processed lines (read from stdin) to the user by rewriting the current line. How to avoid displaying that progress at the top of the terminal on linux?

**Reply by stephentoub** [reply]
> Is there a workaround?

If stdin or stdout is redirected, there's no way to get the current cursor position on Unix.  You can avoid displaying your progress in this fashion if input/output is redirected by checking Console.IsInput/OutputRedirected.
